src/core/matchers/toHaveSpyInteractions.js

```diff
@@ -0,0 +1,73 @@
+getJasmineRequireObj().toHaveSpyInteractions = function(j$) {
+  var getErrorMsg = j$.formatErrorMsg(
+    '<toHaveSpyInteractions>',
+    'expect(<spyObj>).toHaveSpyInteractions()'
+  );
+
+  /**
+   * {@link expect} the actual (a {@link SpyObj}) spies to have been called.
+   * @function
+   * @name matchers#toHaveSpyInteractions
+   * @since 4.0.0
+   * @example
+   * expect(mySpyObj).toHaveSpyInteractions();
+   * expect(mySpyObj).not.toHaveSpyInteractions();
+   */
+  function toHaveSpyInteractions(matchersUtil) {
+    return {
+      compare: function(actual) {
+        var result = {};
+
+        if (!j$.isObject_(actual)) {
+          throw new Error(
+            getErrorMsg(
+              'Expected a spy object, but got ' + matchersUtil.pp(actual) + '.'
+            )
+          );
+        }
+
+        if (arguments.length > 1) {
+          throw new Error(getErrorMsg('Does not take arguments'));
+        }
+
+        result.pass = false;
+        let hasSpy = false;
+        const calledSpies = [];
+        for (const spy of Object.values(actual)) {
+          if (!j$.isSpy(spy)) continue;
+          hasSpy = true;
+
+          if (spy.calls.any()) {
+            result.pass = true;
+            calledSpies.push([spy.and.identity, spy.calls.count()]);
+          }
+        }
+
+        if (!hasSpy) {
+          throw new Error(
+            getErrorMsg(
+              'Expected a spy object with spies, but object has no spies.'
+            )
+          );
+        }
+
+        let resultMessage = 'Expected spy object spies to have been called';
```

Comment:
This doesn't produce the right message when the matcher is negated. GIven this spec:

```
  it('handles negation correctly', function() {
    const foo = jasmine.createSpyObj('foo', ['bar']);
    foo.bar();
    expect(foo).not.toHaveSpyInteractions();
  });
```

I get this message:

```
Expected spy object spies to have been called, the following spies were called: foo.bar called 1 time(s)
```

Something like this would be more appropriate:

```
Expected spy object spies not to have been called, but the following spies were called: foo.bar called 1 time(s)
```

Or better yet:

```
Expected spy object spies not to have been called, but foo.bar was called
```

The most important thing is that the failure message reads correctly in both the normal and negated cases. I'm less opinionated about whether to include the call count, although I think it's going to be excess noise for most users.

Reply:
I would need some guidance here on the inners of Jasmine. How can I detect within toHaveSpyInteractions matcher whether this was called as `expect().toHaveSpyInteractions()` or `expect().not.toHaveSpyInteractions()`? Could you point me to an example or documentation? Thanks!

Reply:
https://jasmine.github.io/tutorials/custom_matcher

Reply:
I had seen the customer_matcher. I think this explanation is what clarified the behaviour:

> Take a look at how other built-in matchers like [toHaveBeenCalledWith](https://github.com/jasmine/jasmine/blob/dfa5b6a53d2b12301b0fbd93d496ab14bacfbf5c/src/core/matchers/toHaveBeenCalledWith.js#L47-L57) do it. They don't detect negation. They just assume that if the matcher passes, the failure message will only be displayed if it was negated.


I've updated the logic and tests now to follow that pattern.